Fix git checkmode on invalid ref

pkingstonxyz opened the pull request:
##### SUMMARY

Fix the output of git checkmode when supplied with a nonexistent ref for the `version` parameter.

Uses `git fetch --dry-run [ref]` to check if the supplied ref actually exists. It requires instantiating a tiny empty git repository, but I do see why this would be undesirable. 

If this approach is unacceptable, I have another branch that can simply issue a warning to the user about check mode (similar to pip check mode in versions that don't support dry run)

Fixes #51580

##### ISSUE TYPE

- Bugfix Pull Request

pkingstonxyz committed:
Add check for is_sha1 and add warnings/failure

- Adds an explicit check for a sha1 hash as the
version in a git module invocation.
- Adds a failure when it can't detect if its a
sha1, remote branch, remote tag, or HEAD.
- Adds a warning to call out to the user that
the git module cannot check sha1's.

pkingstonxyz committed:
Add check for is_sha1 and add warnings/failure

- Adds an explicit check for a sha1 hash as the
version in a git module invocation.
- Adds a failure when it can't detect if its a
sha1, remote branch, remote tag, or HEAD.
- Adds a warning to call out to the user that
the git module cannot check sha1's.

pkingstonxyz committed:
Tweak sha1 detection

- There can be short sha1 strings so I
also check for those
- I also made it a bit more vague because
it's plausible that there could be a false
alarm.

pkingstonxyz committed:
Tweak sha1 detection

- There can be short sha1 strings so I
also check for those
- I also made it a bit more vague because
it's plausible that there could be a false
alarm.

pkingstonxyz committed:
Add integration tests for invalid tags

pkingstonxyz committed:
Add integration tests for invalid tags

pkingstonxyz committed:
Add newline to EOF

pkingstonxyz committed:
Add newline to EOF

pkingstonxyz committed:
Remove unecessary check for checkmode

pkingstonxyz committed:
Remove unecessary check for checkmode

pkingstonxyz committed:
Add comment back for clarity

pkingstonxyz committed:
Add comment back for clarity

pkingstonxyz committed:
Add changelog fragment

pkingstonxyz committed:
Add changelog fragment

pkingstonxyz committed:
Fix pep8 violations

pkingstonxyz committed:
Fix pep8 violations

flowerysong commented on lib/ansible/modules/git.py:
```diff
@@ -785,6 +788,13 @@ def is_not_a_branch(git_path, module, dest):
 
+def looks_like_sha1_string(version: str) -> bool:
+    matches_long = re.match(r"^[a-fA-F0-9]{40}$", version)
+    matches_short = re.match(r"^[a-fA-F0-9]{7}$", version)
```
Why check these two lengths specifically? A commit hash can be as long as 64 characters (when using SHA-256, which will be the default in Git 3) or as short as 4.

pkingstonxyz replied:
Ah this comes from a naive understanding of git objects. I went with these two specifically because the git docs sometimes refer to a 'full, 40-character SHA-1 hash' (https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection). 7 characters because that's the default short hash length on github. I hadn't seen the sha-256 switch in my earlier research. 

I'll convert this PR to a draft and rework this section.

pkingstonxyz committed:
Change to use any number of hex characters

pkingstonxyz committed:
Change to use any number of hex characters

pkingstonxyz committed:
Update to use git fetch --dry-run

pkingstonxyz committed:
Update to use git fetch --dry-run

pkingstonxyz committed:
Add newlines

pkingstonxyz committed:
Add newlines

pkingstonxyz committed:
Fix pep8 violations

pkingstonxyz committed:
Fix pep8 violations

pkingstonxyz committed:
Clean up get_sha_hash function

- Add type hints
- Remove manual path creation so that now git init takes care of it

pkingstonxyz committed:
Clean up get_sha_hash function

- Add type hints
- Remove manual path creation so that now git init takes care of it

pkingstonxyz committed:
Remove dead code

pkingstonxyz committed:
Remove dead code

Shrews commented on lib/ansible/modules/git.py:
```diff
@@ -678,6 +679,20 @@ def get_diff(module, git_path, dest, repo, remote, depth, bare, before, after):
+def get_sha_hash(module: AnsibleModule, git_path: str, remote: str, version: str) -> str:
+    temporary_repository = pathlib.Path(module.tmpdir) / "tmp_repo"
+
+    module.run_command([git_path, 'init'], cwd=temporary_repository)  # Create a bare repo
```
I think you'll need to create the temp dir before calling `run_command()`. The `ignore_invalid_cwd` param defaults to `True` here.

pkingstonxyz replied:
Just added a mkdir here, good catch o7

pkingstonxyz committed:
Create temporary_repository directory

pkingstonxyz committed:
Create temporary_repository directory

pkingstonxyz committed:
Fix spelling

pkingstonxyz committed:
Fix spelling

pkingstonxyz committed:
Merge remote-tracking branch 'origin/fix-git-checkmode-on-invalid-ref-51580' into fix-git-checkmode-on-invalid-ref-51580

pkingstonxyz committed:
Piggyback on previously cloned git repo

pkingstonxyz committed:
Update tests from #85110

Update new tests so that they fail correctly in check/diff mode

pkingstonxyz committed:
Make cwd check more robust, simplify vars
